File: 01_ML/01_supervised/01_regression/01_linear_reg/nodejs/01a_single_var.js
// ML: Linear Regression - Single Variate (Node.js port of ../jupyter/01a_single_var.ipynb)
//
// Predict house price from house area (sq.ft.).
//
//   y = m*x + b
//
// Data:
//   area  | price
//   2600  | 550000
//   3000  | 565000
//   3200  | 610000
//   3600  | 680000
//   4000  | 725000

import MLR from 'ml-regression-multivariate-linear';
import { writePlotlyHtml } from './plot.js';

const data = {
  area:  [2600, 3000, 3200, 3600, 4000],
  price: [550000, 565000, 610000, 680000, 725000],
};
const test = {
  area: [2000, 3300, 4100],
};

console.table(data.area.map((a, i) => ({ area: a, price: data.price[i] })));

// MLR expects 2D arrays for both X and Y (one row per sample).
const X = data.area.map((a) => [a]);
const Y = data.price.map((p) => [p]);

const reg = new MLR(X, Y);

// weights shape: (nFeatures + 1) x nOutputs. Last row is the intercept.
const m = reg.weights[0][0];
const b = reg.weights[reg.weights.length - 1][0];
console.log(`coef (m):     ${m}`);
console.log(`intercept (b): ${b}`);

const Xtest = test.area.map((a) => [a]);
const predictions = reg.predict(Xtest).map((row) => row[0]);

console.table(test.area.map((a, i) => ({ area: a, price: predictions[i] })));

writePlotlyHtml(
  new URL('./01a_single_var_plot.html', import.meta.url).pathname,
  'Linear Regression - Single Variate',
  [
    {
      type: 'scatter',
      mode: 'markers',
      name: 'training',
      x: data.area,
      y: data.price,
      marker: { color: 'magenta', symbol: 'star', size: 12 },
    },
    {
      type: 'scatter',
      mode: 'lines+markers',
      name: 'predicted',
      x: test.area,
      y: predictions,
      line: { color: 'blue' },
    },
  ],
  { xaxis: { title: 'area' }, yaxis: { title: 'price' } },
);

console.log('\nWrote 01a_single_var_plot.html — open it in a browser.');

File: 01_ML/01_supervised/01_regression/01_linear_reg/python/01a_single_var.py
"""ML: Linear Regression - Single Variate (CLI port of ../jupyter/01a_single_var.ipynb).

Predict house price from house area (sq.ft.).

    y = m*x + b

Run:
    python 01a_single_var.py
"""

import matplotlib

matplotlib.use("Agg")  # non-interactive: write PNG instead of opening a window

import matplotlib.pyplot as plt
import pandas as pd
from sklearn import linear_model

PLOT_PATH = "01a_single_var_plot.png"


def main() -> None:
    data = {
        "area":  [2600, 3000, 3200, 3600, 4000],
        "price": [550000, 565000, 610000, 680000, 725000],
    }
    test = {"area": [2000, 3300, 4100]}

    df = pd.DataFrame(data)
    print("training data:")
    print(df.to_string(index=False))

    reg = linear_model.LinearRegression()
    reg.fit(df[["area"]], df["price"])

    print(f"\ncoef (m):      {reg.coef_[0]}")
    print(f"intercept (b): {reg.intercept_}")

    tst = pd.DataFrame(test)
    tst["price"] = reg.predict(tst[["area"]])
    print("\npredictions:")
    print(tst.to_string(index=False))

    plt.figure()
    plt.xlabel("area")
    plt.ylabel("price")
    plt.scatter(df.area, df.price, color="magenta", marker="*", label="training")
    plt.plot(tst.area, tst.price, color="blue", marker="o", label="predicted")
    plt.legend()
    plt.savefig(PLOT_PATH, bbox_inches="tight")
    print(f"\nWrote {PLOT_PATH}")


if __name__ == "__main__":
    main()
